examples/eSealing/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+    /// without iterating through the above array which
+    /// saves execution time in the `register_file` function.
+    witnesses_set: StateSet<AccountAddress, S>,
+}
+
+/// The contract state.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    files: StateMap<String, FileState<S>, S>,
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Creates a new state with no files recorded.
+    fn new(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            files: state_builder.new_map(),
+        }
+    }
+
+    /// Check if a file exists.
+    fn file_exists(&self, file_hash: &str) -> bool {
+        let record = self.files.get(&file_hash.to_string());
+
+        match record {
+            Some(_entry) => true,
+            None => false,
+        }
+    }
+
+    /// Check if a witness exists.
+    fn witness_exists(&self, file_hash: &str, witness: AccountAddress) -> bool {
+        let records = self.files.get(&file_hash.to_string());
+
+        match records {
+            Some(entry) => entry.witnesses_set.contains(&witness),
+            None => false,
+        }
+    }
+
+    /// Get all records from a specific file hash.
+    fn get_records(&self, file_hash: &str) -> ReceiveResult<(Vec<AccountAddress>, Vec<Timestamp>)> {
+        let records = self.files.get(&file_hash.to_string());
+
+        match records {
+            Some(entry) => {
+                // File exists
+                Result::Ok((
+                    entry.witnesses.iter().map(|x| *x.deref()).collect(),
+                    entry.timestamps.iter().map(|x| *x.deref()).collect(),
+                ))
+            }
+            None => Result::Ok(([].to_vec(), [].to_vec())),
+        }
+    }
+
+    /// Add a new records to a specific file hash.
+    fn add_record(&mut self, file_hash: &str, witness: AccountAddress, time: Timestamp) {
+        self.files.get_mut(&file_hash.to_string()).unwrap().timestamps.push(time);
+        self.files.get_mut(&file_hash.to_string()).unwrap().witnesses.push(witness);
+        self.files.get_mut(&file_hash.to_string()).unwrap().witnesses_set.insert(witness);
+    }
+
+    /// Add a new file.
+    fn add_file(&mut self, file_hash: &str, state_builder: &mut StateBuilder<S>) {
+        self.files.insert(file_hash.to_string(), FileState {
+            timestamps:    Vec::new(),
+            witnesses:     Vec::new(),
+            witnesses_set: state_builder.new_set(),
+        });
+    }
+}
+
+/// Tagged events to be serialized for the event log.
+#[derive(Debug, Serial, SchemaType)]
+enum Event {
+    WitnessAdded(WitnessAddedEvent),
+    FileAdded(FileAddedEvent),
+}
+
+/// The WitnessAddedEvent is logged when a new witness is added.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct WitnessAddedEvent {
+    /// Hash of the file to be registered by the witness (sender_account).
+    file_hash: String,
```

Comment:
I would strongly prefer using `[u8; 32]` instead of Strings. The base16 encoding is done in clients.